### PR TITLE
Move `Program` into its own crate and start isolating validation logic

### PR DESCRIPTION
**Pull request by neysofu**
This is the first of a series of PRs that will refactor internal code to better reflect the current state of things after #265. It moves the `struct Program` into its own crate and starts to establish a clear API boundary between two classes of WebAssembly validation rules within SVM:

- smWasm validation, which restricts WebAssembly code to a proper subset (SVM-flavored Wasm);
- and fixed-gas validation, which requires an altogether validation steps and builds a full Control Flow Graph of the code.

Solves https://github.com/spacemeshos/svm/issues/271. Partially solves https://github.com/spacemeshos/svm/issues/272.

### CODE REVIEW

File: crates/gas/src/validation.rs [hunk 1]
```diff
@@ -1,17 +1,13 @@
...
-type ProgramError = crate::ProgramError<FuncIndex>;
+use svm_program::{FuncIndex, Op, Program, ProgramError, ProgramVisitor};
+
+use crate::{CallGraphBuilder, FixedGasError, GraphCycles};
 
-/// Validates a Wasm program.
+/// Further validates a smWasm [`Program`] according to the fixed-gas rules.
 ///
-/// The wasm program is considered INVALID when one of the following:
+/// The smWash [`Program`] is considered INVALID when one of the following:
```
Comment (YaronWittenstein) [inline crates/gas/src/validation.rs:9]: typo `smWash`

File: crates/gas/src/validation.rs [hunk 2]
```diff
@@ -137,77 +129,11 @@ impl ProgramVisitor for ProgramValidator {
                 Ok(())
             }
             _ => {
-                self.validate_non_float(op)?;
-
+                //self.validate_non_float(op)?;
```
Comment (YaronWittenstein) [inline crates/gas/src/validation.rs:132]: can be removed?

File: crates/program/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    /// No valid `svm_alloc` function found.
+    FunctionNotFound { func_name: String },
+    /// Floats not allowed
+    FloatsNotAllowed,
+    /// Too many function imports
+    TooManyFunctionImports,
+    /// Function index is too large
+    FunctionIndexTooLarge,
+    /// `call_indirect` isn't allowed
+    CallIndirectNotAllowed,
```
Comment (YaronWittenstein) [inline crates/program/src/error.rs:19]: This is part of the `Fixed Gas` (disallowing the `call_indirect` opcode)

File: crates/program/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,112 @@
+use parity_wasm::elements::Instruction;
```
Comment (YaronWittenstein) [inline crates/program/src/lib.rs:1]: can you please add:

``rust
#![deny(missing_docs)]
#![deny(unused)]
#![deny(dead_code)]
#![deny(unreachable_code)]
```

File: crates/program/src/program.rs [hunk 5]
```diff
@@ -0,0 +1,125 @@
+use indexmap::IndexMap;
+use parity_wasm::elements::{CodeSection, Module};
+
+use crate::{validate_no_floats, Exports, FuncIndex, Function, Imports, Instruction, ProgramError};
+
+/// A fully parsed and validated smWasm program.
+///
+/// smWash is the language in which Spacemesh smart contracts are written in.
```
Comment (YaronWittenstein) [inline crates/program/src/program.rs:8]: typo: `smWash`

File: crates/program/src/program.rs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+        program.set_imports(imports);
+        program.set_exports(exports);
+
+        validate_no_floats(&program)?;
+        Ok(program)
+    }
+
+    /// Calls [`Program::new`] after parsing `wat_module` and transforming it
+    /// from [WebAssembly Text
```
Comment (YaronWittenstein) [inline crates/program/src/program.rs:60]: Maybe better put the links at the bottom:

```md
[WebAssembly Text Format]

WebAssembly Text Format: https://webassembly.github.io/spec/core/text/index.html)
```

File: crates/program/src/validators.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+impl<F, E> ProgramVisitor for OpcodeValidator<F>
+where
+    F: Fn(&Op) -> Result<(), E>,
+{
+    type Output = ();
+    type Error = E;
+
+    fn on_op(&mut self, op: &Op, _program: &Program) -> Result<(), Self::Error> {
```
Comment (YaronWittenstein) [inline crates/program/src/validators.rs:28]: maybe it worth putting `#[inline]` here

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]
Good job!

File: crates/program/src/error.rs (hunk 3 above)
Comment (neysofu) [reply crates/program/src/error.rs:19]: Ops, missed it!

**Review by YaronWittenstein — APPROVED** [review]

**Pull request merged**